tasks/src/main/scala/dagr/tasks/ScatterGather.scala

```diff
@@ -158,65 +213,88 @@ object ScatterGather {
-      subs.foreach(sub => task ==> sub.chain(task))
-      task
+      subs.foreach { sub => sub.chain(task).foreach { s => task ==> s } }
+      Seq(task)
     }
-
-    override def getTasks: Traversable[_ <: Task] = {
+    override def getTasks: Traversable[_ <: Task] =
       throw new UnsupportedOperationException("getTasks is not supported and should never be called on sub-scatters.")
-    }
   }
 
   /**
     * A type of [[Partitioner]] that is found at stage 2 and beyond where the input is guaranteed to be a task already, and
-    * not part of some domain type.  The given method `f` is used to group the source tasks.
+    * not part of some domain type.
     */
-  private class GroupByPartitioner[Result <: Task, Key](f: Result => Key) extends SimpleInJvmTask with Partitioner[(Key, Seq[Result])] {
-    var partitions: Option[Seq[(Key, Seq[Result])]] = None
-    var sources: Option[Seq[Result]] = None
-    val scatter: Scatter[(Key, Seq[Result])] = Scatter(this)
-    override def run(): Unit = this.sources match {
-      case None => throw new IllegalStateException("GroupByPartitioner.tasks called before _tasks populated.")
-      case Some(_tasks) => this.partitions = Some(_tasks.groupBy(f).toSeq)
+  private trait SecondaryPartitioner[Source <: Task, Result] extends SimpleInJvmTask with Partitioner[Result] {
+    var partitions: Option[Seq[Result]] = None
+    var sources: Option[Seq[Source]] = None
+    val scatter: Scatter[Result] = Scatter(this)
+    protected def toPartitions(sources: Seq[Source]): Seq[Result]
+    override def run(): Unit = {
+      this.sources match {
+        case None => throw new IllegalStateException(s"${getClass.getSimpleName}.sources called before _tasks populated.")
+        case Some(_sources) => this.partitions = Some(toPartitions(_sources))
+      }
     }
   }
+
+  /** A [[SecondaryPartitioner]] whose partitions are the concatenation of the results of one or more [[Scatter]]s. The
+    * given method will ensure that the sources can be converted to [[Scatter]]s. */
+  private class FlatMapPartitioner[Source <: Task, Result](toScatter: Source => Scatter[Result]) extends SecondaryPartitioner[Source, Result] {
+    protected def toPartitions(sources: Seq[Source]): Seq[Result] = sources.map(toScatter).flatMap {
+      case _scatter: SubScatter[Result]        => _scatter.tasks
+      case wrapper: PartitionerWrapper[Result] =>
+        wrapper.partitioner.partitions.getOrElse {
+          throw new IllegalStateException("Scatter/Gather pipeline trying to build before scatters populated.")
+        }
+      case _scatter => throw new IllegalStateException(s"Unknown scatter: ${_scatter.getClass.getSimpleName}")
+    }
+  }
+
+  /** A [[SecondaryPartitioner]] whose partitions are groups of sources grouped by the given method `f`. */
+  private class GroupByPartitioner[Result <: Task, Key](f: Result => Key) extends SimpleInJvmTask with SecondaryPartitioner[Result, (Key, Seq[Result])] {
```

Comment:
To me this is an oxymoron.  In the scatter/gather code:
1. A Partitioner is an operation that takes in a single thing and partitions it into many things
2. The `groupBy` operation is one that takes in many things and groups them into fewer things

Mixing the terminology (and perhaps implementation) here is incredibly confusing.

Reply:
I don't see anything about a `Partitioner` taking one thing, just that it produces a sequence of results (i.e. the partitions).  I think it is completely valid to take a list of things as inputs (ex. `[1,2,3,4,5]`) and produce partitions (ex. `[[1,2],[3,4],[5]]`).  That being said I've tried to rename a few things (i.e. `RePartitioner`) to make it clearer.